### ISSUE TITLE
Very slow prerender on medium sized schema

### ISSUE THREAD

**Issue by Admiral-Enigma** [HiDeoo/starlight-openapi#19]
### Describe the bug

I have a 9.5k line openapi json schema where it takes over 30's pr. prerender of a operation. Seems like there is a lot of unessary copying of objects when prerendering a operation. Seems like its the "collectExternalReferencesInRules" function being called a lot when prerendering - see flamegraph.

![image](https://github.com/HiDeoo/starlight-openapi/assets/13486531/0e2dba26-80f0-4c3e-881b-61e53d1ae6fe)

Long build times
![image](https://github.com/HiDeoo/starlight-openapi/assets/13486531/9cfc7d89-ab6c-42ca-8699-8c3214030212)

### To Reproduce

1. Find a large openapi json schema
2. Run astro build
3. Observe very long build time

### Expected behavior

A operation should take less than 10 seconds to prerender

### How often does this bug happen?

Every time

### System Info

- OS: Windows 11

**Comment by Indyandie** [comment]
I noticed a similar issue using the [openai yml](https://github.com/Indyandie/lucero/blob/main/apis/openai.yml) document, and it generated an error. 

```
 generating static routes 
23:15:33 λ astro-expressive-code/routes/scripts.ts
23:15:33   └─ /_astro/ec.sgewm.js (+3ms)
23:15:33 λ astro-expressive-code/routes/styles.ts
23:15:33   └─ /_astro/ec.usa2f.css (+1ms)
23:15:33 ▶ @astrojs/starlight/404.astro
23:15:33   └─ /404.html (+18ms)
23:15:33 ▶ starlight-openapi/components/Route.astro
23:15:33   ├─ /api/openai/index.html (+528ms)
23:15:34   ├─ /api/openai/operations/listassistants/index.html (+2.89s)
23:15:37   ├─ /api/openai/operations/createassistant/index.html (+4.32s)
23:15:41   ├─ /api/openai/operations/getassistant/index.html (+3.32s)
23:15:44   ├─ /api/openai/operations/modifyassistant/index.html (+4.47s)
23:15:49   ├─ /api/openai/operations/deleteassistant/index.html (+167ms)
23:15:49   ├─ /api/openai/operations/createthread/index.html (+1.68s)
23:15:51   ├─ /api/openai/operations/getthread/index.html (+858ms)
23:15:52   ├─ /api/openai/operations/modifythread/index.html (+989ms)
23:15:53   ├─ /api/openai/operations/deletethread/index.html (+170ms)
23:15:53   ├─ /api/openai/operations/listmessages/index.html (+246ms)
23:15:53   ├─ /api/openai/operations/createmessage/index.html (+4.34s)
23:15:57   ├─ /api/openai/operations/getmessage/index.html (+3.79s)
23:16:01   ├─ /api/openai/operations/modifymessage/index.html (+4.05s)
23:16:05   ├─ /api/openai/operations/createthreadandrun/index.html (+8.48s)
23:16:14   ├─ /api/openai/operations/listruns/index.html (+6.97s)
23:16:21   ├─ /api/openai/operations/createrun/index.html (+8.75s)
23:16:29   ├─ /api/openai/operations/getrun/index.html (+7.15s)
23:16:37   ├─ /api/openai/operations/modifyrun/index.html (+5.93s)
23:16:42   ├─ /api/openai/operations/submittoolouputstorun/index.html (+7.22s)
23:16:50   ├─ /api/openai/operations/cancelrun/index.html (+5.86s)
23:16:56   ├─ /api/openai/operations/listrunsteps/index.html (+228ms)
23:16:56   ├─ /api/openai/operations/getrunstep/index.html (+7.75s)
23:17:03   ├─ /api/openai/operations/listassistantfiles/index.html (+165ms)
23:17:04   ├─ /api/openai/operations/createassistantfile/index.html (+993ms)
23:17:05   ├─ /api/openai/operations/getassistantfile/index.html (+863ms)
23:17:06   ├─ /api/openai/operations/deleteassistantfile/index.html (+309ms)
23:17:06   ├─ /api/openai/operations/listmessagefiles/index.html (+167ms)
23:17:06   ├─ /api/openai/operations/getmessagefile/index.html (+856ms)
23:17:07   ├─ /api/openai/operations/createspeech/index.html (+720ms)
23:17:08   ├─ /api/openai/operations/createtranscription/index.html (+1.15s)
23:17:09   ├─ /api/openai/operations/createtranslation/index.html (+971ms)
23:17:10   ├─ /api/openai/operations/createchatcompletion/index.html (+17.09s)
23:17:27   ├─ /api/openai/operations/createcompletion/index.html

file:///home/buraku/gh/lucero/dist/chunks/prerender_CO0speI1.mjs:4572
    throw new Error(`${prefix}
          ^

Error: Failed to render a `<Code>` component on page https://indyandie.github.io/lucero/api/openai/operations
/createcompletion/:

Missing code to render. The `code` prop must be set to a non-empty string.
    at file:///home/buraku/gh/lucero/dist/chunks/prerender_CO0speI1.mjs:4572:11
    at async AstroComponentInstance.render (file:///home/buraku/gh/lucero/dist/chunks/astro_CvIEmPMv.mjs:1077
:7)
    at async Object.render (file:///home/buraku/gh/lucero/dist/chunks/astro_CvIEmPMv.mjs:1547:7)
    at async renderChild (file:///home/buraku/gh/lucero/dist/chunks/astro_CvIEmPMv.mjs:1028:5) {
  id: 'node_modules/starlight-openapi/components/Route.astro'
}

Node.js v21.2.0
```

**Comment by Admiral-Enigma** [comment]
Problem seems to stem from the nesting of schemas, if I comment out `<Schema nested schemaObject={schema} />` the whole site builds in 4 seconds. 
![image](https://github.com/HiDeoo/starlight-openapi/assets/13486531/c8cb1454-2d9c-486f-af5f-5ca35e493257)

@HiDeoo do you have any idea of how we could make this faster? Maybe somehow integrating Million.js into the library might help

**Comment by HiDeoo** [comment]
Thanks for the report. I'll have to investigate, I'm not even sure without looking at the code where `collectExternalReferencesInRules()` even comes from as I have no recollection of it.

**Comment by HiDeoo** [comment]
The issue should now be fixed in [`v0.6.0`](https://github.com/HiDeoo/starlight-openapi/releases/tag/v0.6.0) that I just published. Here is a small comparison on my machine to build the documentation of the package which include 7 schema.

**Before:**

```
Benchmark 1: pnpm build
  Time (mean ± σ):     82.107 s ±  1.945 s    [User: 95.751 s, System: 5.485 s]
  Range (min … max):   80.345 s … 84.194 s    3 runs
```

**After:**

```
Benchmark 1: pnpm build
  Time (mean ± σ):     11.540 s ±  0.452 s    [User: 17.152 s, System: 1.505 s]
  Range (min … max):   11.161 s … 12.040 s    3 runs
```

---

The issue also mentioned an error related to the `<Code>` component. This was a different issue that is also fixed in the same release.
